lib/archive/diff.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package archive
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	docker "github.com/docker/docker/pkg/archive"
+
+	"github.com/vmware/vic/pkg/trace"
+)
+
+const (
+	// ChangeTypeKey defines the key for the type of diff change stored in the tar Xattrs header
+	ChangeTypeKey = "change_type"
+)
+
+// for sort.Sort
+type changesByPath []docker.Change
+
+func (c changesByPath) Less(i, j int) bool { return c[i].Path < c[j].Path }
+func (c changesByPath) Len() int           { return len(c) }
+func (c changesByPath) Swap(i, j int)      { c[j], c[i] = c[i], c[j] }
+
+// Diff produces a tar archive containing the differences between two filesystems
+func Diff(op trace.Operation, newDir, oldDir string, pathspec map[string]string, data bool) (io.ReadCloser, error) {
+
+	changes, err := docker.ChangesDirs(newDir, oldDir)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Sort(changesByPath(changes))
```

Comment:
I figured it was necessary to insure that directories appear in the archive before the files contained therein, etc. Docker does this (where I got the idea), so it may also be necessary to generate archives that match the sha256 of the same ones generated by docker.